### PR TITLE
pkp/pkp-lib#7265 Refactor editorial decisions and introduce email com…

### PR DESCRIPTION
**Pull request by NateWr**
…poser UI component

- Refactor editorial decisions to use EntityDAO and Repo
- Add class for each editorial decision type
- Add new decision types for back to review, back to copyediting, etc
- Add new decision types for reverting declined decisions
- Add new decision type for skip review
- Add event when decision added
- Add API endpoint to record a decision
- Add step-by-step workflow page to record decisions
- Add steps to send emails, promote files and complete forms
- Add composer UI component with template search, file attachments and multiple locales
- Add support for file attachments by upload, review files, submission files and library files
- Add private API endpoint to get library files
- Add email notification to all authors when editorial decision is made
- Add file genre of submission files to API responses
- Add new email templates for all new email notifications
- Update email templates to use authorSubmissionUrl variable
- Update edit_decision table columns to be nullable
- Update editorial statistics to include skip review as accept decision
- Remove EditDecisionDAO, EditorDecisionActionsManager, and related code

### CODE REVIEW

File: classes/decision/Collector.inc.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
+<?php
+/**
+ * @file classes/decision/Collector.inc.php
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/decision/Collector.inc.php:3]: Did I miss something? It doesn't look like other files in `classes/` are `.php`, and I don't see any `import` statements in this file.

File: classes/decision/Collector.inc.php [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/decision/Collector.inc.php]: Thanks goodness for Stack Overflow! `git log --diff-filter=A --numstat --pretty='COMMIT: %H %cd'` to get all files added in a commit. :)

File: classes/decision/Collector.inc.php [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        $this->submissionIds = $submissionIds;
+        return $this;
+    }
+
+    /**
+     * @copydoc CollectorInterface::getQueryBuilder()
+     */
+    public function getQueryBuilder(): Builder
+    {
+        $qb = DB::table($this->dao->table . ' as ed')
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/decision/Collector.inc.php:109]: Good call!

File: classes/decision/Collector.inc.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (asmecher) [reply classes/decision/Collector.inc.php:3]: I didn't see imports, just a general comment. We currently have both our custom autoloader (which looks for `.inc.php` files based on the namespace) *and* PSR-4-based autoloading support thanks to Composer. In effect this means OJS/OMP/OPS are looking for *both* `.inc.php` and `.php` files when autoloading, but before 3.4.0 gets released we'll remove our custom autoloader (once I get the opportunity to rename everything).

TL;DR: No harm in adding new classes as `.php` rather than `.inc.php`, but if you'd rather not fiddle with it, I'll get it later with the batch rename.

[the remark this answers is not in the document]
Comment (NateWr) [reply classes/decision/Collector.inc.php:3]: Ok, I'll leave it for the batch rename but if you get that in before my merge I can rename these.

File: classes/decision/Decision.inc.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace PKP\decision;
+
+use APP\facades\Repo;
+use Exception;
+use PKP\core\DataObject;
+
+class Decision extends DataObject
+{
+    public const INITIAL_DECLINE = 9;
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/decision/Decision.inc.php:30]: > I would suggest noting here in the code that these were previously the SUBMISSION_EDITOR_DECISION_ constants

I kept the old global constants around at the bottom of this file:

```php
if (!PKP_STRICT_MODE) {
    define('SUBMISSION_EDITOR_DECISION_INITIAL_DECLINE', Decision::INITIAL_DECLINE);
    define('SUBMISSION_EDITOR_DECISION_SKIP_REVIEW', Decision::SKIP_REVIEW);
    define('SUBMISSION_EDITOR_RECOMMEND_ACCEPT', Decision::RECOMMEND_ACCEPT);
    define('SUBMISSION_EDITOR_RECOMMEND_PENDING_REVISIONS', Decision::RECOMMEND_PENDING_REVISIONS);
    define('SUBMISSION_EDITOR_RECOMMEND_RESUBMIT', Decision::RECOMMEND_RESUBMIT);
    define('SUBMISSION_EDITOR_RECOMMEND_DECLINE', Decision::RECOMMEND_DECLINE);
    define('SUBMISSION_EDITOR_DECISION_REVERT_DECLINE', Decision::REVERT_DECLINE);
    define('SUBMISSION_EDITOR_DECISION_REVERT_INITIAL_DECLINE', Decision::REVERT_INITIAL_DECLINE);
    define('SUBMISSION_EDITOR_DECISION_BACK_TO_COPYEDITING', Decision::BACK_TO_COPYEDITING);
    define('SUBMISSION_EDITOR_DECISION_BACK_TO_REVIEW', Decision::BACK_TO_REVIEW);
    define('SUBMISSION_EDITOR_DECISION_BACK_TO_SUBMISSION_FROM_COPYEDITING', Decision::BACK_TO_SUBMISSION_FROM_COPYEDITING);
    define('SUBMISSION_EDITOR_DECISION_SEND_TO_PRODUCTION', Decision::SEND_TO_PRODUCTION);
}
```

Does that look ok or should I add a comment here too?

> I see that there are new constants here, but it doesn't look like these are populated during an upgrade step; what is the relationship between previous decisions and the new set in cases where the constants don't map?

Oh thanks for the reminder! I think that the only two cases that need to be considered are the following:

- ACCEPT changes to SKIP_REVIEW when the decision is in submission stage.
- REVERT_DECLINE changes to REVERT_INITIAL_DECLINE when revert decline decision in submission stage.

I've got these noted down to handle when I write the migrations.

File: classes/decision/Decision.inc.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    public const REVERT_INITIAL_DECLINE = 18;
+    public const BACK_TO_COPYEDITING = 21;
+    public const BACK_TO_REVIEW = 20;
+    public const BACK_TO_SUBMISSION_FROM_COPYEDITING = 22;
+    public const SEND_TO_PRODUCTION = 7;
+
+    /**
+     * Get the decision type for this decision
+     */
+    public function getType(): Type
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/decision/Decision.inc.php:46]: I like to rely on the namespace to provide contextual information in these cases, but you're right this one could get confusing. I'll reword throughout.

File: classes/decision/Workflow.inc.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+use APP\facades\Repo;
+use APP\submission\Submission;
+use PKP\context\Context;
+use PKP\db\DAORegistry;
+use PKP\stageAssignment\StageAssignment;
+use PKP\stageAssignment\StageAssignmentDAO;
+use PKP\submission\reviewRound\ReviewRound;
+use PKP\user\User;
+
+class Workflow
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/decision/Workflow.inc.php:25]: I think of a workflow as any set of steps/tasks that are completed sequentially. For example an author's submission workflow, or an editor's review workflow, etc. The context information is important, though, and that's where the namespace can be important.

When I read this class I read it as "Decision Workflow" because of the namespace. I could change it to `class DecisionWorkflow`, but I guess my sense is that the namespace should provide the context of the class, in order to keep the code less verbose. Like we use `PKP\submission\Repository` instead of `PKP\submission\SubmissionRepository`.

I'm not too committed either way, though. Maybe "workflow" is just the barrier? In the UI Library I've called the components `Steps.vue` and `Step.vue`. I could mirror that and call this `class PKP\decision\Steps`. Would that be better?

File: classes/decision/types/Decline.inc.php [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                    $this->validateNotifyAuthorsAction($action, $actionErrorKey, $validator, $submission);
+                    break;
+                case $this->ACTION_NOTIFY_REVIEWERS:
+                    $this->validateNotifyReviewersAction($action, $actionErrorKey, $validator, $submission, $reviewRoundId);
+                    break;
+            }
+        }
+    }
+
+    public function callback(Decision $decision, Submission $submission, User $editor, Context $context, array $actions)
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/decision/types/Decline.inc.php:105]: Good point. I'm a little nervous about calling it `execute` or a similar `record()` or `recordDecision()`, because it overlaps with `Repo::decision()->add()`. A dev reading the code might expect to be able to do this:

```php
$accept = new Accept();
$accept->execute($decision);
```

In other words, the method is only meant to be called from `Repo::decision()->add()` after that function performs some common actions. It'd be great if I could make it `protected` or something and still invoke it from the repository, but of course that doesn't make sense.

Maybe we could call it `runAdditionalActions()` or something like that?

File: classes/log/SubmissionEmailLogDAO.inc.php [hunk 8]
```diff
@@ -55,6 +60,46 @@ public function getBySubmissionId($submissionId)
...
+     * Get the from or to data as a string
+     *
+     * @param array $addressees Expects Mailable::$to or Mailable::$from
+     */
+    protected function getContactString(array $addressees): string
+    {
+        $contactStrings = [];
+        foreach ($addressees as $addressee) {
+            $contactStrings[] = isset($addressee['name'])
+                ? '"' . $addressee['name'] . '" <' . $addressee['address'] . '>'
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/log/SubmissionEmailLogDAO.inc.php:98]: I might be wrong but I think that it doesn't matter here. This is displayed as text in the email log modal on the author dashboard. I'll double-check this though.

[the remark this answers is not in the document]
Comment (NateWr) [reply classes/log/SubmissionEmailLogDAO.inc.php:98]: Yeah, this is never actually displayed anywhere. It is stored in `email_logs.recipients` as `"Carlo Corino" <ccorino@ccorino.com>`. Should we escape on saving to the database or just when displaying it somewhere?

File: classes/submission/reviewRound/ReviewRoundDAO.inc.php [hunk 9]
```diff
@@ -259,6 +259,25 @@ public function getLastReviewRoundBySubmissionId($submissionId, $stageId = null)
         return $row ? $this->_fromRow($row) : null;
     }
 
+    /**
+     * Check if submission has a review round (in the given stage id)
+     */
+    public function submissionHasReviewRound(int $submissionId, ?int $stageId = null): bool
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/submission/reviewRound/ReviewRoundDAO.inc.php:265]: It seemed wasteful to pull 1-n objects out of the database and instantiate them for a simple `bool` check. But maybe it's not worth it? Happy to go either way here...

File: classes/submissionFile/maps/Schema.inc.php [hunk 10]
```diff
@@ -112,6 +133,27 @@ protected function mapByProperties(array $props, SubmissionFile $item): array
                 continue;
             }
 
+            if ($prop === 'genreIsPrimary') {
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/submissionFile/maps/Schema.inc.php:136]: I've always felt that this should be an enum [primary, supplementary, dependent] instead of two settings. I don't think it's sensible for a file genre to be `isDependent` and `isSupplementary` at the same time, although the configuration allows that. The way that files are actually handled in the article landing page and galley views assumes they are one or the other. Thoughts?

For now I'll adjust the API to return a separate property for the genre which includes `dependent` and `supplementary` info.

File: pages/decision/PKPDecisionHandler.inc.php [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+use PKP\security\authorization\DecisionWritePolicy;
+use PKP\security\authorization\internal\SubmissionRequiredPolicy;
+use PKP\security\Role;
+use PKP\submission\Genre;
+use PKP\submission\GenreDAO;
+use PKP\submission\reviewRound\ReviewRound;
+use PKP\submission\reviewRound\ReviewRoundDAO;
+use Stringy\Stringy;
+
+class PKPDecisionHandler extends Handler
```
[the remark this answers is not in the document]
Comment (NateWr) [reply pages/decision/PKPDecisionHandler.inc.php:37]: Yep, sounds good. I wasn't sure if we'd need an app-level override class yet, but I'll rename this for now.

File: pages/workflow/PKPWorkflowHandler.inc.php [hunk 12]
```diff
@@ -823,6 +840,25 @@ protected function notificationOptionsByStage($user, $stageId, $contextId)
         return false;
     }
 
+    /**
+     * Get a label for a recommendation decision type
+     */
+    protected function getRecommendationLabel(int $decisionType): string
```
[the remark this answers is not in the document]
Comment (NateWr) [reply pages/workflow/PKPWorkflowHandler.inc.php]: Good spot! I needed this to fix a missing email variable in the recommendation messages, too. :+1:

File: classes/submission/reviewRound/ReviewRoundDAO.inc.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (asmecher) [reply classes/submission/reviewRound/ReviewRoundDAO.inc.php:265]: Meh, I probably would've reused the existing call but it's not worth rewriting either way. Please continue!

File: classes/submissionFile/maps/Schema.inc.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (asmecher) [reply classes/submissionFile/maps/Schema.inc.php]: :+1: That's sensible, I do wish we'd gone with an enum rather than two settings. We could revisit that in the future but of course we'd have to consider what happens for users who have already designated files as both.

File: classes/decision/types/Decline.inc.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (asmecher) [reply classes/decision/types/Decline.inc.php:105]: Sure, that's a little lengthy but I'm OK with it.

File: classes/decision/Workflow.inc.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (asmecher) [reply classes/decision/Workflow.inc.php:25]: I have a very slight preference for steps, especially since we're already using it elsewhere. I'm probably reacting to "workflow" because we've had so many discussions about "flexible workflow" and it carries those connotations. So if you think it's worth adapting, please go ahead -- but if you're really OK with it as it stands, then I am too.

[the remark this answers is not in the document]
Comment (NateWr) [reply classes/decision/Workflow.inc.php]: Sounds good, I've changed it to `Steps`.

File: classes/decision/steps/Email.inc.php [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        $this->attachers = $attachers;
+        $this->locales = $locales;
+        $this->mailable = $mailable;
+        $this->recipients = $recipients;
+    }
+
+    /**
+     * Can the editor change the recipients of this email
+     */
+    public function canChangeTo(bool $value): self
```
[the remark this answers is not in the document]
Comment (NateWr) [reply classes/decision/steps/Email.inc.php]: Good call. I've replaced the use of "to" with "recipients" throughout this.

File: controllers/grid/queries/QueriesGridHandler.inc.php [hunk 14]
```diff
@@ -654,28 +677,14 @@ public function updateQuery($args, $request)
...
-                    $currentUser->getId(),
-                    PKPNotification::NOTIFICATION_TYPE_ERROR,
-                    ['contents' => __('email.compose.error')]
-                );
-                trigger_error($e->getMessage(), E_USER_WARNING);
+                $mailable->recipients([$user]);
+
+                Mail::send($mailable);
             }
-            return \PKP\db\DAO::getDataChangedEvent($query->getId());
```
Comment (Vitaliy-1) [inline controllers/grid/queries/QueriesGridHandler.inc.php:677]: @NateWr, this introduced a bug with a page not being refreshed after `Query Form` is saved and `note` not being actually updated. I see that `DAO::getDataChangedEvent` method is deprecated, should it be replaced with a new pattern?

Comment (NateWr) [reply controllers/grid/queries/QueriesGridHandler.inc.php:677]: You can continue to use it for grids.
